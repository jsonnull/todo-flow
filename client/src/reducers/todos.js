// @flow

import type { Todos, Todo, Id, Text, TodosAction } from '../types/todos';

const createTodo = (id: Id, text: Text): Todo => ({
  id,
  text,
  completed: false
});

const toggleTodo = (todos: Todos, id: Id): Todos =>
  todos.map(t => (t.id !== id ? t : { ...t, completed: !t.completed }));

const todos = (state: Todos = [], action: TodosAction): Todos => {
  switch (action.type) {
    case 'ADD_TODO':
      return [...state, createTodo(action.id, action.text)];
    case 'TOGGLE_TODO':
      return toggleTodo(state, action.id);
    default:
      return state;
  }
};

export default todos;
