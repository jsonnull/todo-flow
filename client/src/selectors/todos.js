import { createSelector } from 'reselect';

const todosSelector = state => state.todos;
const visibilityFilterSelector = state => state.visibilityFilter;

export const visibleTodosSelector = createSelector(
  todosSelector,
  visibilityFilterSelector,
  (todos, visibilityFilter) => {
    switch (visibilityFilter) {
      case 'SHOW_COMPLETED':
        return todos.filter(t => t.completed);
      case 'SHOW_ACTIVE':
        return todos.filter(t => !t.completed);
      case 'SHOW_ALL':
      default:
        return todos;
    }
  }
);
