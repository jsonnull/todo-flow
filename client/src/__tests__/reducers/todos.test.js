// @flow

import * as actions from '../../actions';
import todos from '../../reducers/todos';

describe('todos reducer', () => {
  it('should handle ADD_TODO', () => {
    const initialState = [];

    expect(
      todos(initialState, actions.addTodo('Test AddTodo'))
    ).toMatchSnapshot();
  });

  it('should handle TOGGLE_TODO', () => {
    const initialState = [
      {
        text: 'Test AddTodo',
        completed: false,
        id: 0
      }
    ];

    expect(todos(initialState, actions.toggleTodo(0))).toMatchSnapshot();
  });
});
