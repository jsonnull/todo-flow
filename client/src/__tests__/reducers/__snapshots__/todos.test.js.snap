exports[`todos reducer should handle ADD_TODO 1`] = `
Array [
  Object {
    "completed": false,
    "id": 0,
    "text": "Test AddTodo",
  },
]
`;

exports[`todos reducer should handle TOGGLE_TODO 1`] = `
Array [
  Object {
    "completed": true,
    "id": 0,
    "text": "Test AddTodo",
  },
]
`;
